Feature/scatterplotselection rebase2

ResearchDaniel opened the pull request:
Rebase of #829
Enable mouse-drag selection/filtering in scatter plot.

Done: 
- Refactored line rendering and its corresponding settings
- Added BoxSelectionRenderer and corresponding settings (do we want to expose something more here?)
- Now possible to click on empty space to deselect :)
- Use Ctrl for appending selections
- Option for filtering using box selection

Not working:
- Selection rectangle stays when releasing mouse on point in plot


### Appending selections using ctrl
![image](https://user-images.githubusercontent.com/8738140/72228224-34cdcd00-35a5-11ea-82ae-5ddf25cdf16e.png)

### Filtering
![image](https://user-images.githubusercontent.com/8738140/72228245-634ba800-35a5-11ea-8a18-3b495fa1e4ba.png)

### Default options
![image](https://user-images.githubusercontent.com/8738140/72228277-968e3700-35a5-11ea-85ef-7028c39edcd5.png)

martinfalk committed:
PlottingGL: rectangle selection (WIP)

martinfalk committed:
PlottingGL: compile fix

ResearchDaniel committed:
Core/OpenGL: Line rendering refactoring

ResearchDaniel committed:
BaseGL: Refactored LineRenderer into separate class.

ResearchDaniel committed:
PlottingGL: Selection rectangle rendering refactoring

ResearchDaniel committed:
Core/BaseGL/Plotting: Line rendering and plot selection refactoring

Inviwo-Team committed:
Jenkins: Format fixes

martinfalk commented on modules/plottinggl/include/modules/plottinggl/plotters/scatterplotgl.h:
```diff
@@ -177,6 +196,13 @@ class IVW_MODULE_PLOTTINGGL_API ScatterPlotGL {
+    std::optional<std::array<dvec2, 2>> dragRect_;
+
+
+    
```
formatting?

martinfalk commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -417,6 +437,94 @@ auto ScatterPlotGL::addSelectionChangedCallback(std::function<SelectionFunc> cal
+
+void ScatterPlotGL::invokeEvent(Event* event, const ivec2& start, const size2_t& size,
+                                bool useAxisRanges) {
+    if (event->hash() == MouseEvent::chash()) {
```
what about touch events?

ResearchDaniel replied:
Not supported yet. Will have to wait until someone needs it

martinfalk commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -483,15 +592,94 @@ void ScatterPlotGL::objectPicked(PickingEvent* p) {
+        xbuf->dispatch<std::unordered_set<size_t>, dispatching::filter::Scalars>(
+            [min = start[0], max = end[0], isFiltering = isFiltering,
+             selectionCompareOp = selectionCompareOp,
+             filteringCompareOp = filteringCompareOp](auto brprecision) {
```
Use one lambda expression (based on `isFiltering`) instead to avoid if statement inside loop?

ResearchDaniel replied:
Refactored this

martinfalk commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -483,15 +592,94 @@ void ScatterPlotGL::objectPicked(PickingEvent* p) {
+        std::unordered_set<size_t> selectedIndices;
+        for (auto ind : subset) {
+            auto elem = data[ind];
+            if (isFiltering && filteringCompareOp(elem, tmin, tmax)) {
```
same here

ResearchDaniel committed:
BaseGL/Plotting: Line rendering refactoring

ResearchDaniel committed:
Plotting:
- Refactoring for better box selection interaction handling
- Possible to select/filter using mouse in scatter plot

ResearchDaniel committed:
Plotting: Formatting

ResearchDaniel commented:
@petersteneteg Ready for review

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg commented:
how does this selection box relate to a axis selection in the PCP?

petersteneteg commented on include/inviwo/core/datastructures/stipplingsettings.h:
```diff
@@ -0,0 +1,72 @@
+    float offset = 0.f;
+    float worldScale = 4.f;
+    /*
+     * @copydoc StipplingSettingsInterface::getMode
```
StipplingSettingsInterface has no docs to copy...

petersteneteg commented on modules/basegl/include/modules/basegl/properties/linesettingsproperty.h:
```diff
@@ -0,0 +1,88 @@
+class IVW_MODULE_BASEGL_API LineSettingsProperty : public LineSettingsInterface,
+                                                   public CompositeProperty {
+public:
+    LineSettingsProperty(const std::string& identifier, const std::string& displayName,
```
should override clone also...

petersteneteg commented on modules/basegl/include/modules/basegl/rendering/linerenderer.h:
```diff
@@ -0,0 +1,78 @@
+    void setUniforms(Shader& shader, const Mesh& mesh, const Camera& camera, size2_t screenDim);
+    void configureShader(Shader& shader);
+    LineSettings settings_;  //!< Local cache
+    MeshShaderCache lineShaders_;
```
Use the static sharing here?

ResearchDaniel replied:
Can't yet since they also depend on defines.

petersteneteg commented on modules/basegl/src/processors/linerendererprocessor.cpp:
```diff
@@ -0,0 +1,89 @@
+    addPort(outport_);
+    imageInport_.setOptional(true);
+
+    addProperty(lineSettings_);
```
addProperties

JochenJankowai commented:
@petersteneteg I would assume that selections made in the scatterplot directly relate to filtering ranges on the corresponding axes in the PCP **given that there is no other filtering on any other axis in the PCP**.

If there is filtering on other axes in the PCP things get messy because you only see a subspace of the data set in the scatterplot. That means if the user has filtered in the PCP on, say, axis A1 and the scatterplot shows axes A2 and A3, it's arguable whether to filter out the points in the scatterplot according to the PCP or not. It may not be a good idea to link PCPs and scatterplots...

petersteneteg commented on modules/basegl/src/rendering/linerenderer.cpp:
```diff
@@ -0,0 +1,133 @@
+                          const LineSettingsInterface* settings) {
+    if (mesh.getNumberOfBuffers() == 0) return;
+    // Changing these settings require recompilation
+    if (settings_.getPseudoLighting() != settings->getPseudoLighting() ||
```
might extend the MeshShaderCache to handle this state also...

petersteneteg replied:
That would allow sharing the shaders...

ResearchDaniel replied:
Would be nice, future PR :)

petersteneteg commented on include/inviwo/core/datastructures/stipplingsettingsinterface.h:
```diff
@@ -0,0 +1,55 @@
+/*
+ * \brief Settings for stippling (Dashed line, e.g., - - -)
+ */
+class IVW_CORE_API StipplingSettingsInterface {
```
Don't think this belongs in core/datastructues

ResearchDaniel replied:
I don't either but Stippling was already there...
@martinfalk are you using it anywehere else? Where could we put it?

petersteneteg replied:
My guess would be base? @martinfalk?

ResearchDaniel replied:
Moved to base

petersteneteg commented on include/inviwo/core/properties/stipplingproperty.h:
```diff
@@ -32,17 +32,16 @@
 namespace inviwo {
 
-class IVW_CORE_API StipplingProperty : public CompositeProperty {
+class IVW_CORE_API StipplingProperty : public CompositeProperty, public StipplingSettingsInterface {
```
this should probably be in base...

ResearchDaniel replied:
Agree

petersteneteg commented on modules/plotting/src/interaction/boxselectioninteractionhandler.cpp:
```diff
@@ -0,0 +1,205 @@
+                if (static_cast<double>(elem) < min || static_cast<double>(elem) > max) {
+                    continue;
+                } else {
+                    selectedIndices.emplace_back(ind);
```
Does this not need to use the indices of the index buffer in the dataframe? or did we stop doing that? 
Thought the PCP used indices from the indexbuffer?

ResearchDaniel replied:
Hmm, it seem to. Will look into it. Would be good if we at some point could unify this process...

petersteneteg commented on modules/plotting/src/interaction/boxselectioninteractionhandler.cpp:
```diff
@@ -0,0 +1,205 @@
+            [start, end, ybuf = yAxis->getRepresentation<BufferRAM>()](auto brprecision) {
+                using ValueType = util::PrecisionValueType<decltype(brprecision)>;
+                return ybuf->dispatch<std::unordered_set<size_t>, dispatching::filter::Scalars>(
+                    [start, end, xData = brprecision->getDataContainer()](auto brprecision) {
```
think you are making a copy of the container here.

ResearchDaniel replied:
Should return `const &`?

petersteneteg replied:
yes, but `xData = ...`  will make a copy, need to use `&xData = ...` to make a ref...

petersteneteg commented on modules/plotting/src/interaction/boxselectioninteractionhandler.cpp:
```diff
@@ -0,0 +1,205 @@
+                        std::unordered_set<size_t> selectedIndices;
+                        for (auto&& [ind, xVal, yVal] :
+                             util::enumerate(xData, brprecision->getDataContainer())) {
+                            if (static_cast<double>(xVal) < start[0] ||
```
you might what to cast start and end to T outsize of the loop instead... converting to double is quite expensive.

ResearchDaniel replied:
I did so first but then I figured it is not safe because the start/end will be rounded off, which will not produce the same result as casting the value.

petersteneteg replied:
should be possible, but one has to do the rounding a bit more carefully...

ResearchDaniel replied:
Suggestions on how are welcome :)
Simply std::round might work?

ResearchDaniel replied:
Or round up for start and down for end?

petersteneteg replied:
Yes but only in the case that T is not floating point...

ResearchDaniel replied:
Fixed

petersteneteg commented on modules/plotting/src/interaction/boxselectioninteractionhandler.cpp:
```diff
@@ -0,0 +1,205 @@
+            std::copy_if(selectedIndicesX.begin(), selectedIndicesX.end(),
+                         std::inserter(selectedIndices, selectedIndices.begin()),
+                         [data, min, max](auto idx) {
+                             return !(static_cast<double>(data[idx]) < min ||
```
cast min and max outside the loops instead.

petersteneteg commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -285,10 +353,21 @@ void ScatterPlotGL::plot(const size2_t& dims, IndexBuffer* indexBuffer, bool use
+                std::iota(inds.begin(), inds.end(), 0);
+            } else {
+                auto ind = 0;
+                for (auto idx = 0; idx < xbuf->getSize(); idx++) {
```
copy_if?

petersteneteg replied:
make use util::seqeunce for the copy

petersteneteg commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -299,30 +378,48 @@ void ScatterPlotGL::plot(const size2_t& dims, IndexBuffer* indexBuffer, bool use
+                [this](auto val) { return filteredIndices_.find(val) == filteredIndices_.end(); });
+#include <warn/pop>
+
+            if (selectedIndicesGL_.getSize() <
```
This name is very bad... should be getByteSize or something very ugly like that....

petersteneteg commented on modules/plottinggl/src/plotters/scatterplotgl.cpp:
```diff
@@ -464,6 +578,7 @@ void ScatterPlotGL::objectPicked(PickingEvent* p) {
     if (properties_.hovering_.get()) {
         if (p->getHoverState() == PickingHoverState::Enter) {
             hoverIndex_ = id;
+            hoverIndexGL_.upload(static_cast<const void*>(&id), sizeof(uint32_t));
```
Might be bad doing opengl stfuff here.. might not have the right context. That is only garanted in process and initializeResourses, and the constructor.

petersteneteg commented on modules/plottinggl/src/rendering/boxselectionrenderer.cpp:
```diff
@@ -0,0 +1,71 @@
+                                        {0, 1, 2, 3, 0});
+        }
+
+        auto start = vec2((*dragRect)[0]);
```
should use const whenever possible

petersteneteg commented on src/core/datastructures/stipplingsettings.cpp:
```diff
@@ -0,0 +1,52 @@
+ *********************************************************************************/
+
+#include <inviwo/core/datastructures/stipplingsettingsinterface.h>
+#include "..\..\..\include\inviwo\core\datastructures\stipplingsettings.h"
```
fix

ResearchDaniel commented:
@petersteneteg @JochenJankowai 
Selection in scatterplot <-> selection in pcp
Filtering in scatterplot <-> filtering in pcp (ranges are not adjusted)

ResearchDaniel committed:
BaseGL/PlottngGL: Review fixes

Inviwo-Team committed:
Jenkins: Format fixes

Commit:
Stippling include fix

Commit:
LineSettings: Added missing override qualifiers

Commit:
LineSettings: copy method fix

Inviwo-Team committed:
Jenkins: Format fixes

Commit:
BoxSelection: Include fixes

Commit:
ScatterPlotSelection: Jenkins fixes

Inviwo-Team committed:
Jenkins: Format fixes

Commit:
ScatterPlotSelection: Jenkins fixes

ResearchDaniel committed:
Plotting: Make ScatterPlotGL use vector<bool> for selection/filtering

ResearchDaniel committed:
Merge branch 'feature/scatterplotselection-rebase2' of github.com:inviwo/inviwo into feature/scatterplotselection-rebase2

# Conflicts:
#	modules/plotting/src/interaction/boxselectioninteractionhandler.cpp
#	modules/plottinggl/src/plotters/scatterplotgl.cpp
#	modules/plottinggl/src/rendering/boxselectionrenderer.cpp

ResearchDaniel committed:
OpenGL: Rename BufferObject::setSize to BufferObject::setSizeInBytes

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel committed:
Plotting: Bex selection/filtering optimization

ResearchDaniel commented:
@petersteneteg I think that all comments have been addressed now. You might want to take a second look

ResearchDaniel committed:
Merge branch 'feature/scatterplotselection-rebase2' of github.com:inviwo/inviwo into feature/scatterplotselection-rebase2

# Conflicts:
#	modules/plotting/src/interaction/boxselectioninteractionhandler.cpp

ResearchDaniel committed:
OpenCL: Update due to BufferObject change

ResearchDaniel committed:
Plotting: Warn fixes

ResearchDaniel committed:
Core: Moved Stippling settings/property from core to base

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel committed:
Core: Moved Stippling settings/property from core to base

ResearchDaniel committed:
Merge branch 'feature/scatterplotselection-rebase2' of github.com:inviwo/inviwo into feature/scatterplotselection-rebase2

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel committed:
Core: Moved Stippling settings/property from core to base
